### PR TITLE
Warn when building 32-bit macOS applications

### PR DESCRIPTION
**Pull request by chamons**
- https://github.com/xamarin/xamarin-macios/issues/3325

### CODE REVIEW

**Comment by chamons** [comment]
Please bikeshed my warning message hard. I did not know of a great way of writing it, and I am happy for suggestions.

File: docs/website/mmp-errors.md [hunk 1]
```diff
@@ -136,6 +136,14 @@ See [Build optimizations](https://developer.xamarin.com/guides/cross-platform/ma
 
 ### <a name="MM0133"/>MM0133: Found more than 1 assembly matching '{0}' choosing first: '{1}'
 
+### <a name="MM0134"/>MM0134: 32-bit applications should be migrated to 64-bit as they will not run beyond High Sierra "without compromises".
+
+Apple has announced that 32-bit macOS will not run on the version of macOS after High Sierra "without compromises" and will not allow macOS App Store submissions.
```
Comment (spouliot) [inline docs/website/mmp-errors.md]: Suggestions
- split in two
- start with App Store (it already happened) and add a link to Apple announcement
- next "without compromise" again with a link to Apple announcement

File: docs/website/mmp-errors.md [hunk 2]
```diff
@@ -136,6 +136,14 @@ See [Build optimizations](https://developer.xamarin.com/guides/cross-platform/ma
...
+
+In addition, macOS 10.13.4 has added a first launch notification for 32-bit applications.
```
Comment (spouliot) [inline docs/website/mmp-errors.md]: That *first launch notification* sounds weird (but I just might not have the ear for it).

_ macOS 10.13.4 will notify, on the first launch, end users of 32bits (only?) applications_ 

again a link to Apple would be nice

File: docs/website/mmp-errors.md [hunk 3]
```diff
@@ -136,6 +136,14 @@ See [Build optimizations](https://developer.xamarin.com/guides/cross-platform/ma
...
+
+Consider moving your application and any dependencies to Unified 64-bit.
```
Comment (spouliot) [inline docs/website/mmp-errors.md]: moving ? -> updating 
Unified ? not needed, it's the only solution to have 64 bits

File: tests/mmptest/src/MMPTest.cs [hunk 4]
```diff
@@ -240,6 +240,23 @@ public void Unified_HelloWorld_ShouldHaveNoWarnings ()
...
+                };
+
+                string buildResults = TI.TestUnifiedExecutable (test).BuildOutput;
+                Assert.True (buildResults.Contains ("MM0134"), "32-bit Modern did not contain MM0134");
+
+                test.XM45 = true;
+                buildResults = TI.TestUnifiedExecutable (test).BuildOutput;
+                Assert.True (buildResults.Contains ("MM0134"), "32-bit Full did not contain MM0134");
+            });
+        }
```
Comment (spouliot) [inline tests/mmptest/src/MMPTest.cs]: indent seems wrong

Comment (chamons) [reply tests/mmptest/src/MMPTest.cs]: New machine, i thought i hammered VSfM formatting into correct settings but apparently now.

We don't have them somewhere I can apply directly, do we?

File: tools/mmp/driver.cs [hunk 5]
```diff
@@ -1237,6 +1237,9 @@ static int Compile ()
 			if (IsUnified && !arch_set)
 				arch = "x86_64";
 
+			if (arch != "x86_64")
+				ErrorHelper.Warning (134, "32-bit applications should be migrated to 64-bit as they will not run beyond High Sierra \"without compromises\".");
```
Comment (dalexsoto) [inline tools/mmp/driver.cs]: just a friendly reminder to update this one according to Sebastien's comments above :)

File: docs/website/mmp-errors.md (hunk 3 above)
Comment (chamons) [reply docs/website/mmp-errors.md]: Classic XM will see this warning as well. Fair enough though.

File: docs/website/mmp-errors.md [hunk 6]
```diff
@@ -136,6 +136,16 @@ See [Build optimizations](https://developer.xamarin.com/guides/cross-platform/ma
...
 
+### <a name="MM0134"/>MM0134: 32-bit applications should be migrated to 64-bit.
+
+Apple has announced that it will not allow macOS App Store submissions of 32-bit apps (starting January 2018). 
+
+In addition 32-bit applications will not run on the version of macOS after High Sierra "without compromises". 
+
+For more details: https://developer.apple.com/news/?id=06282017a
+
+Consider updating your application and any dependencies to 64-bit.
```
Comment (VincentDondain) [inline docs/website/mmp-errors.md]: And how does one upgrade to 64-bit? (;

Comment (chamons) [reply docs/website/mmp-errors.md]: This is a valid point. Let me link our docs.

Comment (chamons) [reply docs/website/mmp-errors.md]: Let's block this PR until I get the documentation up, so I can link to it.

https://paper.dropbox.com/doc/Updating-Xamarin.Mac-Unified-Applications-to-64-bit-jKVd7srZyC9GoZmD2CMgM

Comment (spouliot) [reply docs/website/mmp-errors.md]: let's not block, we can update the message (several times if needed) once the docs are live

Comment (chamons) [reply docs/website/mmp-errors.md]: Filed https://github.com/xamarin/xamarin-macios/issues/3577 so I don't forget.

**Comment by chamons** [comment]
Test failures known: https://github.com/xamarin/maccore/issues/647 and https://github.com/xamarin/maccore/issues/649

**Pull request merged**